src/tcp_stream.rs

```diff
@@ -166,6 +168,9 @@ impl AsyncRead for TcpStream {
         buf: &mut ReadBuf,
     ) -> Poll<io::Result<()>> {
         let me = unsafe { self.get_unchecked_mut() };
+        if me.is_eof {
```

Comment:
should read out all buffered data before returning EOF, i.e. do this after line 184, I think we can do this check in line  205

Reply:
Agreed. Thank you for confirming.